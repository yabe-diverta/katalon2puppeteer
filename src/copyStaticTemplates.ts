import fs from 'fs';
import glob from 'glob';
import path from 'path';

export function copyStaticTemplates(newDirPath: string) {
  glob
    .sync(path.resolve(__dirname, '..', 'template', '*.js'))
    .forEach((staticTemplatePath) => {
      const dist = path.resolve(newDirPath, path.basename(staticTemplatePath));
      fs.copyFileSync(staticTemplatePath, dist);
    });
}
