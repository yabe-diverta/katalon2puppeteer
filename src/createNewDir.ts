import mkdirp from 'mkdirp';
import path from 'path';
import rimraf from 'rimraf';

export function createNewDir(jsonPath: string) {
  const basename = path.basename(jsonPath);
  const extTrimmedBaseName = basename.replace(path.extname(basename), '');
  const dirName = path.dirname(jsonPath);
  const newDir = path.resolve(dirName, extTrimmedBaseName);

  rimraf.sync(newDir);
  return mkdirp.sync(newDir) || (newDir as string);
}
