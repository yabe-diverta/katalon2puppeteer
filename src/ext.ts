interface String {
  replaceAll(str1: string, str2: string, ignore: boolean): void;
  add(str: string): string;
  encode(): string;
}
interface Number {
  pad(): string;
}

String.prototype.replaceAll = function (str1, str2, ignore) {
  return this.replace(
    new RegExp(
      str1.replace(/([\/\,\!\\\^\$\{\}\[\]\(\)\.\*\+\?\|\<\>\-\&])/g, '\\$&'),
      ignore ? 'gi' : 'g'
    ),
    typeof str2 == 'string' ? str2.replace(/\$/g, '$$$$') : str2
  );
};

String.prototype.add = function (str = '') {
  return `${this}${str}`;
};

String.prototype.encode = function () {
  return encodeURIComponent(`${this}`);
};

Number.prototype.pad = function () {
  return `000${this}`.slice(-3);
};
