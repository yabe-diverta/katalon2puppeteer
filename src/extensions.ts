interface String {
  replaceAll(str1: string, str2: string, ignore: boolean): void;
  add(str: string): string;
}
interface Number {
  pad(): string;
}

String.prototype.replaceAll = function (str1, str2, ignore) {
  return this.replace(
    new RegExp(
      str1.replace(/([\/\,\!\\\^\$\{\}\[\]\(\)\.\*\+\?\|\<\>\-\&])/g, '\\$&'),
      ignore ? 'gi' : 'g'
    ),
    typeof str2 == 'string' ? str2.replace(/\$/g, '$$$$') : str2
  );
};

String.prototype.add = function (str = '') {
  return `${this}${str}`;
};

Number.prototype.pad = function () {
  return `000${this}`.slice(-3);
};
