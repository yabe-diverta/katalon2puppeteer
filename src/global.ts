import { CommandOption } from './type';

export class Global {
  static option: CommandOption;
  static JSONs: string[];
}
