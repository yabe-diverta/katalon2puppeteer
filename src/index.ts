import './ext';
import glob from 'glob';
import { copyStaticTemplates } from './copyStaticTemplates';
import { createNewDir } from './createNewDir';
import { TaskBuilder } from './TaskBuilder';
import { CommandOption } from './type';

export function create(option: CommandOption) {
  glob.sync(option.input).forEach((jsonPath) => {
    const newDirPath = createNewDir(jsonPath);
    copyStaticTemplates(newDirPath);
    new TaskBuilder()
      .init(jsonPath, newDirPath)
      .createTaskFileDefinitions()
      .createIndexJsDefinition(option)
      .writeFiles()
  });
}
