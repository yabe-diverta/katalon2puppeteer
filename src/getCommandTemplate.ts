import mkdirp from 'mkdirp';
import path from 'path';
import { Global } from './global';
import { transpile } from './transpliler';
import { PuppeteerJson, PuppeteerJsonObj } from './type';

const getCapturePath = (
  fileIdx: number,
  idx: number,
  def: PuppeteerJsonObj
) => {
  const id = Object.entries(def)
    .map(([k, v]) => `${k}=${v}`)
    .join(',');
  const specFilePath = Global.JSONs[fileIdx];
  const fileName = `${path.basename(specFilePath)}_${idx.pad()}_${id}`
    .encode()
    .add('.png');

  const dirPath =
    Global.option.captureDir !== undefined
      ? path.join(Global.option.captureDir, path.dirname(specFilePath))
      : path.join(path.dirname(specFilePath), 'capture');
  mkdirp(dirPath);

  return path.join(path.relative(process.cwd(), dirPath), fileName);
};

export function getCommandTemplate(defs: PuppeteerJson, fileIdx: number) {
  return defs
    .map(transpile)
    .map(({ code, def }, idx) =>
      code
        .add(
          `
            await delay(${Global.option.delay})`
        )
        .add(
          Global.option.capture
            ? `
            await page.screenshot({
              path: "${getCapturePath(fileIdx, idx, def)}",
              type: 'png',
              fullPage: true
            });`
            : ``
        )
    )
    .map((code) =>
      code
        .split('\n')
        .map((line) => line.trim())
        .filter((line) => line !== '')
        .map((line) => `\t${line}`)
        .join('\n')
    )
    .join('\n\n');
}
