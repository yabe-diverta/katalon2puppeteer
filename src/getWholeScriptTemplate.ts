import path from 'path';
import { Global } from './global';

export function getWholeScriptTemplate(commandTemplate: string, idx: number) {
  return `
const fs = require('fs')
const puppeteer = require('puppeteer');

function mkdir(dirName) {
  if (!fs.existsSync(dirName)){
    fs.mkdirSync(dirName);
  }
}
mkdir('${path.join(path.dirname(Global.JSONs[idx]), 'capture')}');

function delay(time) {
  return new Promise((resolve) => setTimeout(resolve, time));
}

(async () => {
    const browser = await puppeteer.launch({
      headless: ${Global.option.headless},
      defaultViewport: {
        width: ${Global.option.viewportWidth},
        height: ${Global.option.viewportHeight}
      },
      args: ['--start-maximized']
    });
    const page = await browser.newPage();
      try {
      const extensions = require('puppeteer-extensions')(page);
      extensions.turnOffAnimations();
    } catch (e) {
      const msg = \`
      \\x1b[33mpuppeteer-extensions would work better for your e2e testing,
      see https://github.com/HuddleEng/puppeteer-extensions#miscellaneous
      \\x1b[0m\${e}
      \`
        .split('\\n')
        .map((line) => line.trim())
        .join('\\n');
      console.info(msg);
    }
    let element, formElement, tabs;

${commandTemplate}

    await browser.close();
})();
`;
}
