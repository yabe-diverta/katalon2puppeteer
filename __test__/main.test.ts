import fs from 'fs';
import * as Transpiler from '../src';

describe('snapshot main transpilation', () => {
  test.skip('will match snapshot', () => {
    // arrange
    fs.mkdirSync = jest.fn().mockReturnValue(() => {});
    fs.writeFileSync = jest.fn().mockReturnValue(() => {});

    // act
    Transpiler.create({
      input: './__test__/etc/test.json',
      delay: 500,
    });

    // assert
    const outDir = (fs.mkdirSync as jest.Mock).mock.calls[0][0];
    const outFile = (fs.writeFileSync as jest.Mock).mock.calls[0][0];
    const code = (fs.writeFileSync as jest.Mock).mock.calls[0][1];

    // console.dir(outDir);
    // console.dir(outFile);
    // console.dir(code);

    expect(fs.mkdirSync).toBeCalled();
    expect(fs.writeFileSync).toBeCalled();

    expect(outDir).toBe(`${process.cwd()}/__test__/etc`);
    expect(outFile).toBe(`./__test__/etc/test.e2e.test.js`);
    expect(code).toMatchSnapshot();
  });
});
