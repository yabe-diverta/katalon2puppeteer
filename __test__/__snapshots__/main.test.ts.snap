// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot main transpilation will match snapshot 1`] = `
"
const fs = require('fs')
const puppeteer = require('puppeteer');

function mkdir(dirName) {
  if (!fs.existsSync(dirName)){
    fs.mkdirSync(dirName);
  }
}
const captureDir = process.argv.slice(2)[0] !== undefined
  ? process.argv.slice(2)[0]
  : '__test__/etc/capture';
mkdir(captureDir);

function delay(time) {
  return new Promise((resolve) => setTimeout(resolve, time));
}

(async () => {
  try {
    const browser = await puppeteer.launch({
      headless:
        process.argv.slice(2)[1] !== undefined &&
        process.argv.slice(2)[1] === '--headless',
      defaultViewport: {
        width: 1920,
        height: 1080
      },
      args: ['--start-maximized', '--lang=en-US']
    });
  
    const page = await browser.newPage();
    await page.evaluateOnNewDocument(() => {
      Object.defineProperty(navigator, 'language', {
        get: () => 'en-US',
      });
      Object.defineProperty(navigator, 'languages', {
        get: () => ['en-US', 'en'],
      });
    });
    await page.setExtraHTTPHeaders({
      'Accept-Language': 'en-US',
    });
  
    try {
      const extensions = require('puppeteer-extensions')(page);
      extensions.turnOffAnimations();
    } catch (e) {
      const msg = \`
      \\\\x1b[33mpuppeteer-extensions would work better for your e2e testing,
      see https://github.com/HuddleEng/puppeteer-extensions#miscellaneous
      \\\\x1b[0m\${e}
      \`
        .split('\\\\n')
        .map((line) => line.trim())
        .join('\\\\n');
      console.info(msg);
    }
    let element, formElement, tabs;
  

  await page.authenticate({
    username: 'userId',
    password: 'password',
  });
  
	await page.goto(\\"http://test.example.com:8080/\\", { waitUntil: 'networkidle0' });
	await delay(500)
	await page.screenshot({
	path: \`\${captureDir}\${require('path').sep}test.json_000_command%3Dopen%2Ctarget%3Dhttp%3A%2F%2Ftest.example.com%3A8080%2F%2Cvalue%3D.png\`,
	type: 'png',
	fullPage: true
	});
	console.log('puppeteer executed at no.0')

	await page.waitForXPath(\\"//a[contains(text(),'link1')]\\")
	element = await page.$x(\\"//a[contains(text(),'link1')]\\");
	await element[0].click().catch(async (e) => await page.evaluate((elm) => { elm.click(); }, element[0]))
	await page.waitForNavigation();
	await delay(500)
	await page.screenshot({
	path: \`\${captureDir}\${require('path').sep}test.json_001_command%3Dclick%2Ctarget%3Dlink%3Dlink1%2Cvalue%3Dbin.png\`,
	type: 'png',
	fullPage: true
	});
	console.log('puppeteer executed at no.1')

	await page.waitForXPath(\\"//img[@alt='image']\\")
	element = await page.$x(\\"//img[@alt='image']\\");
	await element[0].click().catch(async (e) => await page.evaluate((elm) => { elm.click(); }, element[0]))
	await delay(500)
	await page.screenshot({
	path: \`\${captureDir}\${require('path').sep}test.json_002_command%3Dclick%2Ctarget%3D%2F%2Fimg%5B%40alt%3D'image'%5D%2Cvalue%3D.png\`,
	type: 'png',
	fullPage: true
	});
	console.log('puppeteer executed at no.2')

	await page.waitForXPath(\\"//p[2]/span\\")
	element = await page.$x(\\"//p[2]/span\\");
	await element[0].click().catch(async (e) => await page.evaluate((elm) => { elm.click(); }, element[0]))
	await delay(500)
	await page.screenshot({
	path: \`\${captureDir}\${require('path').sep}test.json_003_command%3Dclick%2Ctarget%3D%2F%2Fp%5B2%5D%2Fspan%2Cvalue%3D.png\`,
	type: 'png',
	fullPage: true
	});
	console.log('puppeteer executed at no.3')
  
    await browser.close();
  } catch(e) {
    console.error(e);
    process.exit(1);
  }
})();
"
`;
